File: auth0_guardian_ios/ios/Classes/GuardianDeviceFlutter.swift
import Flutter
import Foundation
import Guardian

public class GuardianDeviceFlutter: NSObject {
  static let METHOD_CHANNEL_NAME = "com.kpler/auth0_guardian_ios/device"

  var plugin: SwiftFlutterPlugin?

  init(plugin: SwiftFlutterPlugin) {
    super.init()
    self.plugin = plugin
    let channel = FlutterMethodChannel(
      name: GuardianDeviceFlutter.METHOD_CHANNEL_NAME,
      binaryMessenger: plugin.registrar!.messenger()
    )
    channel.setMethodCallHandler(handle(_:result:))
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    let arguments = call.arguments as? NSDictionary
    let args = call.arguments as! [String: Any]

    switch call.method {
    case "delete":
      delete(args: args, result: result)
    case "update":
      update(args: args, result: result)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  public func delete(args: [String: Any], result: @escaping FlutterResult) {
    do {
      // Extract the arguments from the method call (Flutter).
      let forDomain = args["forDomain"] as! String
      let forEnrollmentId = args["forEnrollmentId"] as! String
      let userId = args["userId"] as! String

      // Get the signing key and the verification key.
      let signingKey = try SigningKeyService(domain: forDomain).getSigningKey()

      // Trigger the deletion.
      Guardian
        .api(forDomain: forDomain)
        .device(forEnrollmentId: forEnrollmentId, userId: userId, signingKey: signingKey)
        .delete()
        .start { response in
          switch response {
          case .success:
            result(true)
          case .failure(let error):
            result(
              getGuardianFlutterError(error, "DeviceFlutter.delete() deleting failed")
            )
          }
        }
    } catch (let error) {
      result(
        getGuardianFlutterError(error, "DeviceFlutter.delete() throwed an unexpected error")
      )
    }
  }

  public func update(args: [String: Any], result: @escaping FlutterResult) {
    do {
      // Extract the arguments from the method call (Flutter).
      let forDomain = args["forDomain"] as! String
      let forEnrollmentId = args["forEnrollmentId"] as! String
      let userId = args["userId"] as! String

      let localIdentifier = args["localIdentifier"] as? String? ?? nil
      let name = args["name"] as? String? ?? nil
      let notificationToken = args["notificationToken"] as? String? ?? nil

      // Get the signing key and the verification key.
      let signingKey = try SigningKeyService(domain: forDomain).getSigningKey()

      // Trigger the deletion.
      Guardian
        .api(forDomain: forDomain)
        .device(forEnrollmentId: forEnrollmentId, userId: userId, signingKey: signingKey)
        .update(localIdentifier: localIdentifier, name: name, notificationToken: notificationToken)
        .start { response in
          switch response {
          case .success:
            result(true)
          case .failure(let error):
            result(
              getGuardianFlutterError(error, "DeviceFlutter.update() updating failed")
            )
          }
        }
    } catch (let error) {
      result(
        getGuardianFlutterError(error, "DeviceFlutter.update() throwed an unexpected error")
      )
    }
  }

  public func dispose() {
    plugin = nil
  }

  deinit {
    dispose()
  }
}

File: auth0_guardian_android/android/src/main/kotlin/com/kpler/auth0_guardian_flutter/GuardianDeviceFlutter.kt
package com.kpler.auth0_guardian_flutter

import EnrolledDevice
import android.net.Uri
import com.auth0.android.guardian.sdk.Enrollment
import com.auth0.android.guardian.sdk.Guardian
import com.auth0.android.guardian.sdk.networking.Callback
import com.kpler.auth0_guardian_flutter.utils.SigningKeyService
import com.kpler.auth0_guardian_flutter.utils.getGuardianErrorMessage
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler

class GuardianDeviceFlutter : MethodCallHandler {
    private val methodChannelName = "com.kpler/auth0_guardian_android/device"

    private var channel: MethodChannel

    private var plugin: KotlinFlutterPlugin?

    constructor(plugin: KotlinFlutterPlugin) {
        this.plugin = plugin
        this.channel = MethodChannel(plugin.flutterPluginBinding.binaryMessenger, methodChannelName)
        this.channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        val args = call.arguments as Map<String, Any>
        when (call.method) {
            "delete" -> delete(args, result)
            else -> result.notImplemented()
        }
    }

    private fun buildGuardian(domain: String): Guardian {
        val url = Uri.parse(domain)
        return Guardian.Builder().url(url).build()
    }

    private fun delete(args: Map<String, Any>, result: MethodChannel.Result) {
        try {
            // Extract the arguments from the method call (Flutter).
            val forDomain = args["forDomain"] as String
            val forEnrollmentId = args["forEnrollmentId"] as String
            val userId = args["userId"] as String

            // Get the signing key.
            val signingKey = SigningKeyService(forDomain).getSigningKey()
            val enrollment: Enrollment = EnrolledDevice(signingKey, userId, forEnrollmentId)

            // Define the callback to handle the enrollment result.
            val callback =
                    object : Callback<Void> {
                        override fun onSuccess(repsonse: Void?) {
                            SigningKeyService(forDomain).deleteSigningKey()
                            result.success(true)
                        }
                        override fun onFailure(exception: Throwable) {
                            result.error(
                                    "DeleteFailure",
                                    getGuardianErrorMessage(
                                            "GuardianDeviceFlutter.delete() failed"
                                    ),
                                    exception.message
                            )
                        }
                    }

            // Start the enrollment process.
            buildGuardian(forDomain).delete(enrollment).start(callback)
        } catch (exception: Exception) {
            result.error(
                    "UnexpectedError",
                    getGuardianErrorMessage(
                            "GuardianDeviceFlutter.delete() thrown an unexpected error"
                    ),
                    exception.message
            )
        }
    }

    fun dispose() {
        channel.setMethodCallHandler(null)
        plugin = null
    }
}
